tests/Unit/Helpers/Utilities/Serialization/Versioning.hpp

```diff
@@ -0,0 +1,140 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include "Framework/TestingFramework.hpp"
+
+#include <algorithm>
+#include <cstddef>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "Utilities/Serialization/Serialize.hpp"
+#include "Utilities/TypeTraits/IsA.hpp"
+
+namespace TestHelpers::serialization {
+namespace versioning_detail {
+std::vector<std::pair<std::string, std::vector<std::byte>>> read_serializations(
+    const std::string& filename);
+
+void write_serialization(const std::string& filename, const std::string& label,
+                         const std::vector<std::byte>& serialization);
+
+template <typename Compare>
+struct ToCompare {
+  const Compare& operator()(const Compare& x) const { return x; }
+
+  template <typename Base>
+  const Compare& operator()(const std::unique_ptr<Base>& x) const {
+    return dynamic_cast<const Compare&>(*x);
+  }
+};
+
+template <>
+struct ToCompare<void> {
+  template <typename Serialize>
+  const Serialize& operator()(const Serialize& x) const {
+    static_assert(not tt::is_a_v<std::unique_ptr, Serialize>,
+                  "Cannot compare unique_ptrs.  Pass a template argument to "
+                  "test_versioning to set type to compare.");
+    return x;
+  }
+};
+}  // namespace versioning_detail
+
+/// Test serialization of a versioned class against old versions.
+///
+/// Old serializations of the class are stored in \p filename
+/// (relative to the unit test source directory).  Each stored version
+/// has a label, where the current label must match \p current_label.
+/// This function checks that
+///
+/// * The last entry matches \p current_label and matches the result
+///   of serializing \p current_object.
+///
+/// * Old lines in the file are successfully deserialized and give
+///   expected values.
```

Comment:
Ah I see how this works now. While an interesting idea, I suggest to simplify this function to just deserialize the data with the given label and check equality with the given object. That makes the unit tests easier to understand, gives the flexibility to test variations of objects, and allows handling of newer options. The `.serializations` file is just a dictionary of serializations then.

Reply:
I think the current design allows all that, and additionally ensures that you can't forget to test any entries and that the file is up-to-date.

Reply:
I agree with @wthrowe . I think to do what you want you could just have `current_label` be an older version, then set up the objects as you want in the map. The only thing that's currently not clear to me on how we want to handle is the case where deserialization of an old version is just not possible and we throw an `ERROR`. Could you add to the dox how to test that an appropriate error is thrown in that case?

Reply:
Hadn't thought of the ERROR case.  I've made the old versions into a variant so you can check for error messages instead of success.

If you want to test multiple sets of parameters you can call the function multiple times with different file names.

Reply:
Ok